src/execution/operator/join/physical_piecewise_merge_join.cpp

```diff
@@ -147,6 +177,73 @@ unique_ptr<LocalSinkState> PhysicalPiecewiseMergeJoin::GetLocalSinkState(Executi
 	return move(result);
 }
 
+static idx_t PiecewiseMergeNulls(DataChunk &keys, const vector<JoinCondition> &conditions) {
+	// Merge the validity masks of the comparison keys into the primary
+	// Return the number of NULLs in the resulting chunk
+	D_ASSERT(keys.ColumnCount() > 0);
+	const auto count = keys.size();
+
+	size_t all_constant = 0;
+	for (auto &v : keys.data) {
+		all_constant += int(v.GetVectorType() == VectorType::CONSTANT_VECTOR);
+	}
+
+	auto &primary = keys.data[0];
+	if (all_constant == keys.data.size()) {
+		//	Either all NULL or no NULLs
+		for (auto &v : keys.data) {
+			if (ConstantVector::IsNull(v)) {
+				ConstantVector::SetNull(primary, true);
+				return count;
+			}
+		}
+		return 0;
+	} else if (keys.ColumnCount() > 1) {
+		//	Normalify the primary, as it will need to merge arbitrary validity masks
+		primary.Normalify(count);
+		auto &pvalidity = FlatVector::Validity(primary);
+		for (size_t c = 1; c < keys.data.size(); ++c) {
+			// Skip comparisons that accept NULLs
+			if (conditions[c].comparison == ExpressionType::COMPARE_DISTINCT_FROM) {
+				continue;
+			}
+			//	Orrify the rest, as the sort code will do this anyway.
+			auto &v = keys.data[c];
+			VectorData vdata;
+			v.Orrify(count, vdata);
+			auto &vvalidity = vdata.validity;
+			if (vvalidity.AllValid()) {
+				continue;
+			}
+			pvalidity.EnsureWritable();
+			auto pmask = pvalidity.GetData();
+			if (v.GetVectorType() == VectorType::FLAT_VECTOR) {
```

Comment:
We need to Orrify because the keys are computed from an input chunk. But there is nothing in that branch that references the vector, so the test just needs to be removed.

Reply:
Correct me if I'm wrong but the way I'm reading the code below is that it requires the validity mask to hold consecutive values (i.e. it requires the vector to be a flat vector). The `VectorData` struct + selection vector needs to be used to correctly support different vector types.

Perhaps the current code can stay as an optimization for flat vectors (which does not require `Orrify` to be called), and the Orrify can be moved as a back-up strategy for other vector types?

Reply:
Oh right the masks are tied to the selection vector.